platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/isolated/IsolatedProjectsCompositeBuildIntegrationTest.groovy

```diff
@@ -50,40 +48,98 @@ class IsolatedProjectsCompositeBuildIntegrationTest extends AbstractIsolatedProj
         fixture.assertStateLoaded()
     }
 
-    def "cycles for plugin builds are prohibited"() {
+    def "'root(lib) -> plugins -> root(lib)' is allowed"() {
```

Comment:
❌ This notation is confusing to me. I suggest using prose, e.g., `def "cycle between root build and plugin build is allowed"()` but I'd be ok with a less confusing notation.

Why is this cycle allowed?

Reply:
@alllex @bamboo Does the answer above make sense for you?

Reply:
The current implementation short-circuits the inclusion process, deferring the substitution registration. It's hard to see why the root build is special in this case, given that the library build cycles are supported in general.

However, one thing the short-circuiting does not take into account is whether the inclusion of root happens as a plugin build. I believe this is broken by definition: it should not be possible to include the root build as a plugin build. I wonder if it works today in any meaningful way. But this is something we can definitely forbid with IP.

Reply:
> it should not be possible to include the root build as a plugin build.

➕